### PR TITLE
Me/formatter

### PR DESCRIPTION
**Pull request by meysholdt**

### CODE REVIEW

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeEObjectRegion.java [hunk 1]
```diff
@@ -47,4 +48,11 @@ public AbstractElement getGrammarElement() {
 		return null;
 	}
 
+	@Override
+	public String getText() {
+		if (node instanceof RootNode)
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeEObjectRegion.java]: Please don't use implementation classes. This looks like you want to do node.getRootNode().getText()

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractTextSegment.java [hunk 2]
```diff
@@ -61,20 +61,31 @@ public int getEndOffset() {
 	}
 
 	@Override
-	public ITextSegment getIndentation() {
-		return getTextRegionAccess().indentationRegion(getOffset());
+	public int getLineCount() {
+		return Strings.countLines(getText());
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractTextSegment.java]: I think you may want to use org.eclipse.xtext.util.Strings.countLineBreaks(CharSequence) instead

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	protected NodeEObjectRegion createTokens(NodeModelBasedRegionAccess access, INode node) {
+		return new NodeEObjectRegion(access, node);
+	}
+
+	protected Map<EObject, AbstractEObjectRegion> getEObjectToTokensMap(ITextRegionAccess regionAccess) {
+		this.eObjToTokens = Maps.newHashMap();
+		this.firstHidden = createHiddenRegion(regionAccess);
+		this.lastHidden = this.firstHidden;
+		NodeModelBasedRegionAccess access = (NodeModelBasedRegionAccess) regionAccess;
+		CompositeNode rootNode = (CompositeNode) resource.getParseResult().getRootNode();
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java]: Please avoid using impl classes if not strictly necessary.

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	protected void process(INode node, NodeModelBasedRegionAccess access) {
+		NodeEObjectRegion tokens = stack.peek();
+		boolean creator = isEObjectRoot(node);
+		if (creator || tokens == null) {
+			tokens = new NodeEObjectRegion(access, node);
+			tokens.setLeadingHiddenRegion(lastHidden);
+			stack.push(tokens);
+		}
+		if (tokens.getSemanticElement() == null) {
+			if (node instanceof RootNode)
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java]: node.getParent() == null is equivalent

**Comment by meysholdt** [comment]
Jenkins test

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractTextSegment.java (hunk 2 above)
Comment (meysholdt) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractTextSegment.java]: fixed in 3b56ebd3d0858f5224cebd408c6b3f6fb576b856

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java (hunk 3 above)
Comment (meysholdt) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java]: fixed in 3e1c696e86a9d51ddc244af82b94cbce2b79318c

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java (hunk 4 above)
Comment (meysholdt) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccessBuilder.java]: fixed in 3e1c696e86a9d51ddc244af82b94cbce2b79318c

**Comment by meysholdt** [comment]
Jenkins test

**Comment by meysholdt** [comment]
Jenkins test

**Comment by meysholdt** [comment]
Jenkins test

**Comment by meysholdt** [comment]
Jenkins test

**Pull request merged**